core/src/main/scala/scalaz/IList.scala

```diff
@@ -293,11 +292,21 @@ sealed abstract class IList[A] extends Product with Serializable {
   def reduceRightOption(f: (A, A) => A): Option[A] =
     reverse.reduceLeftOption((a, b) => f(b, a))
 
-  def reverse: IList[A] =
-    foldLeft(IList.empty[A])((as, a) => a :: as)
+  def reverse: IList[A] = {
+    @tailrec def go(as: IList[A], acc: IList[A]): IList[A] = as match {
+      case c : ICons[_] => go(c.tail, c.head :: acc)
```

Comment:
Vanilla `::` is a constructor, while this `::` here is a method, if memory serves. Is there any performance to be gained by calling the `ICons` constructor directly?

Reply:
with `ICons`...

```
[info] Benchmark                 Mode  Cnt   Score   Error  Units
[info] MapBench.ilist100         avgt    5   0.825 ± 0.007  us/op
[info] MapBench.ilist1000        avgt    5   8.396 ± 0.052  us/op
[info] MapBench.ilist10000       avgt    5  75.159 ± 8.505  us/op
[info] MapBench.ilistClass100    avgt    5   0.864 ± 0.009  us/op
[info] MapBench.ilistClass1000   avgt    5   9.033 ± 0.138  us/op
[info] MapBench.ilistClass10000  avgt    5  84.445 ± 1.009  us/op
```

Reply:
This looks pretty optimal to me unless you want to change the representation and special case a bunch of stuff (for example, a list of length 100 can be mapped on the stack).